src/fsharp/FSharp.Core/array.fs

```diff
@@ -494,12 +494,23 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Filter")>]
         let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = List<_>() // ResizeArray
```

Comment:
@KevinRansom I think your comment addresses the code he's replacing. In his new implementation, he only ever allocates 2 arrays and there is no reallocation.